### PR TITLE
Add SERIAL_SENSOR support (#2477)

### PR DESCRIPTION
**Pull request by pbek**
Implements https://github.com/xoseperez/espurna/issues/2477

### CODE REVIEW

File: code/espurna/serialsensor.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    terminalRegisterCommand(F("SENSOR0"), [](const terminal::CommandContext& ctx) {
+        if (ctx.argc == 2) {
+            String value = String(ctx.argv[1]);
+            mqttSend("SENSOR0", value.c_str());
+            ctx.output.printf_P(PSTR("SENSOR0 has value %s\n"),
+                value.c_str());
+        }
+
+        terminalOK(ctx);
+    });
```
Comment (mcspr) [inline code/espurna/serialsensor.cpp]: b/c of the code repetition, maybe what we want is `mqtt.send` command with 2 args - 'topic' and 'payload'? could also just add it to the mqtt.cpp directly
also note of the 'retained' option, should these be ephemeral messages?

File: code/espurna/serialsensor.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+// -----------------------------------------------------------------------------
+// COMMUNICATIONS
+// -----------------------------------------------------------------------------
+
+/**
+ * Setup function
+ *
+ * I haven't managed to get the terminalRegisterCommand commands in a method
+ * because of the lambda function
```
Comment (mcspr) [inline code/espurna/serialsensor.cpp]: btw there's at least one way to 'proxy' certain dynamic arguments inside of a lambda like this:
https://github.com/xoseperez/espurna/blob/62d83c6194630076fa198ddbf0a899ae7fb79672/code/espurna/sensors/PZEM004TSensor.h#L414-L415
(and true, assuming that this only gets called exactly once)

File: code/espurna/serialsensor.cpp (hunk 1 above)
Comment (pbek) [reply code/espurna/serialsensor.cpp]: > b/c of the code repetition, maybe what we want is mqtt.send command with 2 args - 'topic' and 'payload'? could also just add it to the mqtt.cpp directly

But how would I get the variables into the callback of `terminalRegisterCommand`? I couldn't sneak them in in the `[]`.

> also note of the 'retained' option, should these be ephemeral messages?

Yes, I guess they could be retained.

File: code/espurna/serialsensor.cpp (hunk 2 above)
Comment (pbek) [reply code/espurna/serialsensor.cpp]: Sneaky, I'll try that. 😁  Thank you.

File: code/espurna/serialsensor.cpp (hunk 1 above)
Comment (mcspr) [reply code/espurna/serialsensor.cpp]: > But how would I get the variables into the callback of terminalRegisterCommand? I couldn't sneak them in in the [].

Which ones though? I meant something like this

```cpp
terminalRegisterCommand(F("MQTT.SEND"), [](const terminal::CommandContext& ctx) {
    if (ctx.argc == 3) {
        if (mqttSend(ctx.argv[1].c_str(), ctx.argv[2].c_str(), false, false)) {
            terminalOK(ctx);
        } else {
            terminalError(ctx, F("Cannot queue the message")); //
        }
        return;
    }

    terminalError(ctx, F("MQTT.SEND <topic> <payload>"));
});
```

Unless we somehow want the command names to end up in the `commands` / `help`

Comment (pbek) [reply code/espurna/serialsensor.cpp]: Yes, sounds like a great idea too. I'll try that later this week!

Comment (pbek) [reply code/espurna/serialsensor.cpp]: In now, works nice.

Comment (mcspr) [reply code/espurna/serialsensor.cpp]: > In now, works nice.

Thanks! Will pull things tomorrow

**Comment by mcspr** [comment]
Sorry for the delay :)
I have tried to use [Codespaces](https://github.dev/pbek/espurna/blob/feature/serial-sensor/code/espurna/main.cpp) to fixup based on my last comment, but it failed to push to the writable PR branch... So that involved some manual git manipulations with a proper PC.

I moved the func to the mqtt.cpp, so no need for the extra `-DSERIAL_SENSOR_SUPPORT=1` and it just works when MQTT is enabled.

**Comment by pbek** [comment]
Great too, thank you very much! 😁

**Pull request merged**